moldesign/utils/classes.py

```diff
@@ -29,6 +34,67 @@ def add(self, item):
         self[key].append(item)
 
 
+class ExclusiveList(object):
+    """ Behaves like a list, but won't allow duplicate items with duplicate keys to be added.
```

Comment:
Yeah, this is basically halfway between a `set` and a `list` - it exposes a List-like interface; but it raises an exception if you try to pass in a duplicate object; and it automatically computes key functions to check if objects are "equivalent" (i.e., `key(o1) == key(o2)`. In sets/dicts, objects are equivalent if `o1.__hash__()==o2.__hash__() and o1 == o2`)

I ended up doing it this way to add some extra validation to an already-existing `list` object (`Molecule.constraints`) without breaking the list interface. Biggest potential issue is that object access to random items in an `ExclusiveList` is O(N) - this probably isn't performance critical, but it may need to be fixed if it becomes a problem.